app/lib/service/email/backend.dart

```diff
@@ -0,0 +1,175 @@
+import '../../shared/datastore.dart';
+import '../../shared/email.dart';
+import 'models.dart';
+
+final _logger = Logger('email.backend');
+final _random = Random.secure();
+
+/// Sets the email backend service.
+void registerEmailBackend(EmailBackend backend) =>
+    ss.register(#_emailBackend, backend);
+
+/// The active email backend service.
+EmailBackend get emailBackend => ss.lookup(#_emailBackend) as EmailBackend;
+
+/// Represents the backend for the outgoing email queue.
+class EmailBackend {
+  final DatastoreDB _db;
+
+  EmailBackend(this._db);
+
+  /// Creates [OutgoingEmail] entities that can be stored alongside a transaction.
+  List<OutgoingEmail> prepareEntities(EmailMessage msg) {
+    return msg.recipients
+        .map(
+          (recipient) => OutgoingEmail.init(
+            fromEmail: msg.from.email,
+            recipientEmail: recipient.email,
+            subject: msg.subject,
+            bodyText: msg.bodyText,
+          ),
+        )
+        .toList();
+  }
+
+  /// Queries all [OutgoingEmail] objects and tries to send out the email,
+  /// deleting the entry after the email was sent successfully. This method
+  /// should be called only from the background task.
+  ///
+  /// The processing of the Datastore query will be stopped after [stopAfter]
+  /// duration has elapsed. This allows the periodic task to complete within
+  /// the planned time window.
+  ///
+  /// When [ignorePendingAt] is true, the newly created entries are
+  /// sent immediately. This is intended for testing only to force-send
+  /// queued entries.
+  ///
+  /// Returns the number of successfully sent emails.
+  Future<int> trySendAllOutgoingEmails({
+    Duration? stopAfter,
+    @visibleForTesting bool ignorePendingAt = false,
+  }) async {
+    final sw = Stopwatch()..start();
+    final query = _db.query<OutgoingEmail>()..order('-created');
+    var successful = 0;
+    await for (final m in query.run()) {
+      if (stopAfter != null && sw.elapsed > stopAfter) break;
+      if (m.isNotAlive) continue;
+      if (!ignorePendingAt && !m.mayAttemptNow) continue;
+      final wasSent =
+          await _trySendOutgoingEmail(m.uuid, ignorePendingAt: ignorePendingAt);
+      if (wasSent) {
+        successful++;
+      }
+    }
+    return successful;
+  }
+
+  /// Tries to send emails with the given [ids]. The
+  /// [OutgoingEmail] entry will be deleted after the
+  /// email was sent successfully.
+  ///
+  /// This method should be called right after the entries
+  /// are saved in the Datastore.
+  Future<void> trySendOutgoingEmails(Iterable<String> ids) async {
+    for (final id in ids) {
+      await _trySendOutgoingEmail(id, ignorePendingAt: true);
+    }
+  }
+
+  /// Tries to send email with the given [id]. The
+  /// [OutgoingEmail] entry will be deleted after the
+  /// email was sent successfully.
+  ///
+  /// Returns true if the email was sent successfully.
+  Future<bool> _trySendOutgoingEmail(
```

Comment:
If instead of taking `id` we took `OutgoingEmail`, then we don't need `ignorePendingAt`.

Instead we skip sending if `OutgoingEmail.pendingAt` has changed since the instance passed in `OutgoingEmail`.

This way, by setting `OutgoingEmail.pendingAt` into the future, we'll have locked the row. Thus, there can be no race between the background process re-attempting to send emails, and the initial method invocation that tries to send emails.

Indeed, even if we ran two background processes at the same time, they would never try to send the same email.

Reply:
Done.

Note: I think we can't really rule out double sending the same email if we have two background processes running, but we are using now two independent locks that should minimize the risk of that.